rosbag2_py/src/rosbag2_py/_transport.cpp

```diff
@@ -201,13 +204,19 @@ class Recorder
     recorder->record();
     rclcpp::executors::SingleThreadedExecutor exec;
     exec.add_node(recorder);
-    auto spin_thread = std::thread(
-      [&exec]() {
-        exec.spin();
-      });
 
-    exec.cancel();
-    spin_thread.join();
+    cancel_ = std::promise<void>();
+    auto future = cancel_.get_future();
+    // Release the GIL for long-running record, so that calling Python code can use other threads
+    {
+      py::gil_scoped_release release;
+      exec.spin_until_future_complete(future);
+    }
+  }
+
+  void cancel()
+  {
+    cancel_.set_value();
```

Comment:
I was able to reproduce the problem locally as well - It's not clear why setting the value on the promise does not wake the executor. However, did we already rule out the following? 

Make `exec_` a SingleThreadedExecutor member variable, and do the following (could get rid of `cancel_`)

```suggestion
    // Release the GIL for long-running record, so that calling Python code can use other threads
    {
      py::gil_scoped_release release;
      exec_.spin();
    }
  }

  void cancel()
  {
    exec_.cancel();
```